Composer/packages/client/src/recoilModel/undo/history.ts

```diff
@@ -1,18 +1,154 @@
+    const { dialogId, selected, focused } = undoHistory.present === 0 ? undoHistory.initialLocation : location.contents;
+    const currentUri = `/bot/${projectId.contents}/dialogs/${dialogId}${getUrlSearch(selected, focused)}`;
+    navigateTo(currentUri);
+  }
 }
 
-const undoHistory = new UndoHistory();
-export default undoHistory;
+function mapTrackedAtomsOntoSnapshot(current: Snapshot, target: Snapshot): Snapshot {
+  const assets = getAssets(current);
+  trackedAtoms.forEach((atom, index) => {
+    const loadable = assets[index];
+    if (loadable.state === 'hasValue') {
+      target = target.map(({ set }) => set(atom, loadable.contents));
+    }
+  });
+  return target;
+}
+
+function setInitialLocation(snapshot: Snapshot, undoHistory: UndoHistory) {
+  const location = snapshot.getLoadable(designPageLocationState);
+  if (location.state === 'hasValue') {
+    undoHistory.setInitialLocation({ ...location.contents });
+  }
+}
+
+export const UndoRoot = React.memo(() => {
+  const history = useRef(undoHistory).current;
+  const setUndoFunction = useSetRecoilState(undoFunctionState);
+  const [, forceUpdate] = useState([]);
+  //use to record the first time change, this will help to get the init location
+  //init location is used to undo navigate
+  const assetsChanged = useRef(false);
+
+  useRecoilTransactionObserver(({ snapshot, previousSnapshot }) => {
+    if (checkAtomChanged(snapshot, previousSnapshot, projectIdState)) {
+      //switch project should clean the undo history
+      undoHistory.clear();
+      undoHistory.add(snapshot);
+    } else if (!assetsChanged.current) {
+      if (checkAtomsChanged(snapshot, previousSnapshot, trackedAtoms)) {
+        assetsChanged.current = true;
+      }
+      setInitialLocation(snapshot, history);
+    }
+  });
+
+  const undoAssets = (target: Snapshot, current: Snapshot, gotoSnapshot: (snapshot: Snapshot) => void) => {
+    target = target.map(({ set }) => set(isUndoingState, true));
+    target = mapTrackedAtomsOntoSnapshot(current, target);
+    gotoSnapshot(target);
+    navigate(current, history);
+  };
+
+  const undo = useRecoilCallback(({ snapshot, gotoSnapshot }: CallbackInterface) => () => {
+    if (history.canUndo()) {
+      const current = history.undo();
+      undoAssets(snapshot, current, gotoSnapshot);
+    }
+  });
+
+  const redo = useRecoilCallback(({ snapshot, gotoSnapshot }: CallbackInterface) => () => {
+    if (history.canRedo()) {
+      const current = history.redo();
+      undoAssets(snapshot, current, gotoSnapshot);
+    }
+  });
+
+  const canUndo = () => {
+    return history.canUndo();
+  };
+
+  const canRedo = () => {
+    return history.canRedo();
+  };
+
+  const commit = useRecoilCallback(({ snapshot }) => () => {
+    const previousSnapshot = history.getPresentSnapshot();
+    //filter some invalid changes
+    if (previousSnapshot && checkAtomsChanged(snapshot, previousSnapshot, trackedAtoms)) {
+      history.add(snapshot);
```

Comment:
Are entire snapshots of the application saved in the state? Like if we go through a 100 actions without undoing, redoing is every snapshot of the whole app saved in the stack? Lets connect offline on this file for better understanding @lei9444